File: ios/Runner/ShareHandler.swift
import Foundation
import Flutter
import UIKit

class ShareHandler {
    private var methodChannel: FlutterMethodChannel?
    private var initialShareData: [String: Any?]?
    
    init(methodChannel: FlutterMethodChannel?) {
        self.methodChannel = methodChannel
    }
    
    func getInitialShareData() -> [String: Any?] {
        let data = initialShareData ?? [:]
        initialShareData = nil // Clear setelah diambil
        return data
    }
    
    func handleSharedContent(url: URL) {
        // Handle URL dari share extension atau file
        let fileManager = FileManager.default
        
        if fileManager.fileExists(atPath: url.path) {
            let fileInfo = processFile(at: url)
            
            let data: [String: Any?] = [
                "type": "file",
                "text": nil,
                "subject": nil,
                "files": [fileInfo]
            ]
            
            sendToFlutter(data: data)
        }
    }
    
    private func processFile(at url: URL) -> [String: Any?] {
        let fileName = url.lastPathComponent
        let filePath = url.path
        
        // Get MIME type
        let mimeType = FileUtils.getMimeType(for: url)
        let fileType = categorizeFileType(mimeType: mimeType)
        
        // Get file size
        let fileSize = FileUtils.getFileSize(at: url)
        let sizeFormatted = fileSize != nil ? FileUtils.formatFileSize(bytes: fileSize!) : nil
        
        // Generate thumbnail
        var thumbnailPath: String?
        switch fileType {
        case "image":
            thumbnailPath = FileUtils.generateImageThumbnail(from: url, size: 512)
        case "video":
            thumbnailPath = FileUtils.generateVideoThumbnail(from: url, size: 512)
        default:
            thumbnailPath = nil
        }
        
        return [
            "path": filePath,
            "uri": url.absoluteString,
            "name": fileName,
            "type": fileType,
            "mimeType": mimeType as Any,
            "thumbnail": thumbnailPath as Any,
            "size": fileSize as Any,
            "sizeFormatted": sizeFormatted as Any
        ]
    }
    
    private func categorizeFileType(mimeType: String?) -> String {
        guard let mimeType = mimeType else { return "file" }
        
        switch true {
        case mimeType.hasPrefix("image/"):
            return "image"
        case mimeType.hasPrefix("video/"):
            return "video"
        case mimeType.hasPrefix("audio/"):
            return "audio"
        case mimeType == "application/pdf":
            return "pdf"
        case mimeType.contains("wordprocessing") || mimeType.contains("msword"):
            return "document"
        case mimeType.contains("spreadsheet") || mimeType.contains("excel"):
            return "spreadsheet"
        case mimeType.contains("presentation") || mimeType.contains("powerpoint"):
            return "presentation"
        case mimeType.contains("zip") || mimeType.contains("compressed"):
            return "archive"
        case mimeType.hasPrefix("text/"):
            return "text"
        default:
            return "file"
        }
    }
    
    private func sendToFlutter(data: [String: Any?]) {
        if let channel = methodChannel {
            // App sudah running, kirim langsung
            channel.invokeMethod("onShareReceived", arguments: data)
        } else {
            // App baru dibuka, simpan untuk nanti
            initialShareData = data
        }
    }
}

File: android/app/src/main/kotlin/id/bee/sharemanual/ShareHandler.kt
package id.bee.sharemanual

import android.content.ContentResolver
import android.content.Intent
import android.net.Uri
import io.flutter.plugin.common.MethodChannel
import java.io.File

class ShareHandler(
    private val contentResolver: ContentResolver,
    private val cacheDir: File,
    private val methodChannel: MethodChannel?
) {
    
    private var initialShareData: HashMap<String, Any?>? = null
    
    fun handleIntent(intent: Intent?) {
        if (intent == null) return

        val action = intent.action
        val type = intent.type

        when (action) {
            Intent.ACTION_SEND -> {
                if (type != null) {
                    when {
                        type.startsWith("text/") -> handleSendText(intent)
                        else -> handleSendFile(intent, "file")
                    }
                }
            }
            Intent.ACTION_SEND_MULTIPLE -> {
                if (type != null) {
                    handleSendMultipleFiles(intent, type)
                }
            }
        }
    }
    
    fun getInitialShareData(): HashMap<String, Any?> {
        val data = initialShareData ?: hashMapOf()
        initialShareData = null
        return data
    }
    
    private fun handleSendText(intent: Intent) {
        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
        val sharedSubject = intent.getStringExtra(Intent.EXTRA_SUBJECT)
        
        val data = hashMapOf<String, Any?>(
            "type" to "text",
            "text" to sharedText,
            "subject" to sharedSubject,
            "files" to null
        )
        
        sendToFlutter(data)
    }
    
    private fun handleSendFile(intent: Intent, fileType: String) {
        val fileUri: Uri? = intent.getParcelableExtra(Intent.EXTRA_STREAM)
        
        if (fileUri != null) {
            val filePath = FileUtils.copyUriToCache(fileUri, cacheDir, contentResolver)
            val fileName = FileUtils.getFileName(fileUri, contentResolver)
            val fileSize = FileUtils.getFileSize(fileUri, contentResolver)
            
            val actualMimeType = FileUtils.getMimeType(fileUri, contentResolver)
            val detectedType = categorizeFileType(actualMimeType)
            
            // Generate thumbnail
            val thumbnailPath = when (detectedType) {
                "image" -> FileUtils.generateImageThumbnail(
                    fileUri, 
                    cacheDir, 
                    contentResolver,
                    512
                )
                "video" -> {
                    if (filePath != null) {
                        FileUtils.generateVideoThumbnail(filePath, cacheDir, 512)
                    } else {
                        null
                    }
                }
                else -> null
            }
            
            // Format file size dengan null check
            val sizeFormatted = fileSize?.let { FileUtils.formatFileSize(it) }
            
            val fileInfo = hashMapOf<String, Any?>(
                "path" to filePath,
                "uri" to fileUri.toString(),
                "name" to fileName,
                "type" to detectedType,
                "mimeType" to actualMimeType,
                "thumbnail" to thumbnailPath,
                "size" to fileSize,
                "sizeFormatted" to sizeFormatted
            )
            
            val data = hashMapOf<String, Any?>(
                "type" to "file",
                "text" to null,
                "files" to listOf(fileInfo)
            )
            
            sendToFlutter(data)
        }
    }
    
    private fun handleSendMultipleFiles(intent: Intent, mimeType: String) {
        val fileUris: ArrayList<Uri>? = intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM)
        
        if (fileUris != null && fileUris.isNotEmpty()) {
            val files = fileUris.mapNotNull { uri ->
                val filePath = FileUtils.copyUriToCache(uri, cacheDir, contentResolver)
                if (filePath != null) {
                    val actualMimeType = FileUtils.getMimeType(uri, contentResolver)
                    val fileType = categorizeFileType(actualMimeType)
                    val fileSize = FileUtils.getFileSize(uri, contentResolver)
                    
                    // Generate thumbnail
                    val thumbnailPath = when (fileType) {
                        "image" -> FileUtils.generateImageThumbnail(
                            uri,
                            cacheDir,
                            contentResolver,
                            512
                        )
                        "video" -> FileUtils.generateVideoThumbnail(
                            filePath,
                            cacheDir,
                            512
                        )
                        else -> null
                    }
                    
                    // Format file size dengan null check
                    val sizeFormatted = fileSize?.let { FileUtils.formatFileSize(it) }
                    
                    hashMapOf<String, Any?>(
                        "path" to filePath,
                        "uri" to uri.toString(),
                        "name" to FileUtils.getFileName(uri, contentResolver),
                        "type" to fileType,
                        "mimeType" to actualMimeType,
                        "thumbnail" to thumbnailPath,
                        "size" to fileSize,
                        "sizeFormatted" to sizeFormatted
                    )
                } else {
                    null
                }
            }
            
            val data = hashMapOf<String, Any?>(
                "type" to "files",
                "text" to null,
                "files" to files
            )
            
            sendToFlutter(data)
        }
    }
    
    private fun categorizeFileType(mimeType: String?): String {
        return when {
            mimeType == null -> "file"
            mimeType.startsWith("image/") -> "image"
            mimeType.startsWith("video/") -> "video"
            mimeType.startsWith("audio/") -> "audio"
            mimeType == "application/pdf" -> "pdf"
            mimeType == "application/msword" -> "document"
            mimeType == "application/vnd.openxmlformats-officedocument.wordprocessingml.document" -> "document"
            mimeType == "application/vnd.ms-excel" -> "spreadsheet"
            mimeType == "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" -> "spreadsheet"
            mimeType == "application/vnd.ms-powerpoint" -> "presentation"
            mimeType == "application/vnd.openxmlformats-officedocument.presentationml.presentation" -> "presentation"
            mimeType == "application/zip" -> "archive"
            mimeType == "application/x-rar-compressed" -> "archive"
            mimeType == "application/x-7z-compressed" -> "archive"
            mimeType == "application/vnd.android.package-archive" -> "apk"
            mimeType.startsWith("text/") -> "text"
            mimeType.startsWith("application/") -> "file"
            else -> "file"
        }
    }
    
    private fun sendToFlutter(data: HashMap<String, Any?>) {
        if (methodChannel != null) {
            methodChannel.invokeMethod("onShareReceived", data)
        } else {
            initialShareData = data
        }
    }
}